pkg/manager/member/tiflash_member_manager.go

```diff
@@ -0,0 +1,858 @@
+	svcLister                       corelisters.ServiceLister
+	podLister                       corelisters.PodLister
+	nodeLister                      corelisters.NodeLister
+	autoFailover                    bool
+	tiflashFailover                 Failover
+	tiflashScaler                   Scaler
+	tiflashUpgrader                 Upgrader
+	tiflashStatefulSetIsUpgradingFn func(corelisters.PodLister, pdapi.PDControlInterface, *apps.StatefulSet, *v1alpha1.TidbCluster) (bool, error)
+}
+
+// NewTiFlashMemberManager returns a *tiflashMemberManager
+func NewTiFlashMemberManager(
+	pdControl pdapi.PDControlInterface,
+	setControl controller.StatefulSetControlInterface,
+	svcControl controller.ServiceControlInterface,
+	certControl controller.CertControlInterface,
+	typedControl controller.TypedControlInterface,
+	setLister v1.StatefulSetLister,
+	svcLister corelisters.ServiceLister,
+	podLister corelisters.PodLister,
+	nodeLister corelisters.NodeLister,
+	autoFailover bool,
+	tiflashFailover Failover,
+	tiflashScaler Scaler,
+	tiflashUpgrader Upgrader) manager.Manager {
+	kvmm := tiflashMemberManager{
+		pdControl:       pdControl,
+		podLister:       podLister,
+		nodeLister:      nodeLister,
+		setControl:      setControl,
+		svcControl:      svcControl,
+		certControl:     certControl,
+		typedControl:    typedControl,
+		setLister:       setLister,
+		svcLister:       svcLister,
+		autoFailover:    autoFailover,
+		tiflashFailover: tiflashFailover,
+		tiflashScaler:   tiflashScaler,
+		tiflashUpgrader: tiflashUpgrader,
+	}
+	kvmm.tiflashStatefulSetIsUpgradingFn = tiflashStatefulSetIsUpgrading
+	return &kvmm
+}
+
+// Sync fulfills the manager.Manager interface
+func (tfmm *tiflashMemberManager) Sync(tc *v1alpha1.TidbCluster) error {
+	if tc.Spec.TiFlash == nil {
+		return nil
+	}
+
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+
+	if !tc.PDIsAvailable() {
+		return controller.RequeueErrorf("TidbCluster: [%s/%s], waiting for PD cluster running", ns, tcName)
+	}
+
+	// Sync TiFlash Headless Service
+	if err := tfmm.syncHeadlessService(tc); err != nil {
+		return err
+	}
+
+	return tfmm.syncStatefulSet(tc)
+}
+
+func (tfmm *tiflashMemberManager) syncHeadlessService(tc *v1alpha1.TidbCluster) error {
+	if tc.Spec.Paused {
+		klog.V(4).Infof("tiflash cluster %s/%s is paused, skip syncing for tiflash service", tc.GetNamespace(), tc.GetName())
+		return nil
+	}
+
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+
+	newSvc := getNewHeadlessService(tc)
+	oldSvcTmp, err := tfmm.svcLister.Services(ns).Get(controller.TiFlashPeerMemberName(tcName))
+	if errors.IsNotFound(err) {
+		err = controller.SetServiceLastAppliedConfigAnnotation(newSvc)
+		if err != nil {
+			return err
+		}
+		return tfmm.svcControl.CreateService(tc, newSvc)
+	}
+	if err != nil {
+		return err
+	}
+
+	oldSvc := oldSvcTmp.DeepCopy()
+
+	equal, err := controller.ServiceEqual(newSvc, oldSvc)
+	if err != nil {
+		return err
+	}
+	if !equal {
+		svc := *oldSvc
+		svc.Spec = newSvc.Spec
```

Comment:
```suggestion
		svc.Spec = newSvc.Spec
		svc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
```
?

Reply:
This is a headless service, no need to preserve the clusterIP.